Elytra in overworld

babbaj opened the pull request:
<!-- No UwU's or OwO's allowed -->
Started by @underscore-zi 
nether-pathfinder changes have been merged in https://github.com/babbaj/nether-pathfinder/pull/20
Closes #4187

underscore-zi committed:
Add initial support for overworld/end flight

babbaj committed:
update to nether-pathfinder api/concurrency changes

underscore-zi committed:
Make usage of the Baritone cache optional for elytra flight

underscore-zi committed:
Set a better default Y for elytra GoalXZ

underscore-zi committed:
Make sure player is in bounds before attempting elytra pathfinding

underscore-zi committed:
Allow landing on bedrock (necessary for nether roof)

underscore-zi committed:
Revert "Set a better default Y for elytra GoalXZ"

This reverts commit 984c71e01d3406289ba0a6031da4ad0200ec75e1.

underscore-zi committed:
GoalXZ y-level will consider if player is above nether roof

underscore-zi committed:
Rework landing to be safed when descending from high y-levels, and utilize height map to save some cycles if player is flying above ground.

underscore-zi committed:
Check negative y landing spots too

underscore-zi committed:
Merge branch 'cabaletta:1.19.4' into 1.19.4

babbaj committed:
dynamically change destination y if the roof gets in the way

babbaj committed:
1.6

babbaj committed:
memset if no air also cave air is real

thelampgod reviewed (approved):
wow!

babbaj committed:
custom allocator by default :^)

underscore-zi committed:
Respect allowAboveRoof roof setting when looking for safe landing spots

underscore-zi committed:
Allow landing on cobblestone and obsidian

underscore-zi commented:
I do have a couple minor commits over on https://github.com/underscore-zi/baritone 

First is respecting the allowRoof setting when searching for a landing spot. I had left a TODO in for this but hadn't done so yet.

Second is adding Cobblestone and Obsidian to allowed landing blocks. In vanilla the current list is fine but I was using this around 2b's spawn and it had some issues finding landing spots.

ZacSharp commented:
Can you please wait three days with merging? I'd like to properly read through and test this, but won't be able to do so before April 8.

babbaj commented:
leijurv never merges my PRs anyways

leijurv commented:
> Can you please wait three days with merging?

say no more, i'll check back next month

ZacSharp commented on src/api/java/baritone/api/Settings.java:
```diff
@@ -1543,6 +1543,26 @@ public final class Settings {
+    /**
+     * Allow the pathfinder to access the baritone cache to improve pathing
+     */
+    public final Setting<Boolean> elytraUseBaritoneCache = new Setting<>(true);
```
This setting name / documentation is odd. From a user perspective it's all Baritone.

ZacSharp commented on src/main/java/baritone/process/elytra/BlockStateOctreeInterface.java:
```diff
@@ -39,16 +39,16 @@ public BlockStateOctreeInterface(final NetherPathfinderContext context) {
 
     public boolean get0(final int x, final int y, final int z) {
-        if ((y | (127 - y)) < 0) {
+        if (y < 0 || y > 383) {
```
This always checks for the maximum chunk height, is that correct? i.e. does nether-pathfinder always allocate full-size chunks?

babbaj replied:
yes

ZacSharp commented on src/main/java/baritone/process/elytra/NetherPathfinderContext.java:
```diff
@@ -24,98 +24,153 @@
+    public NetherPathfinderContext(long seed, Path cache, Level world) {
+        this.dimension = world.dimension();
+        final int dim;
+        if (this.dimension == Level.NETHER) dim = NetherPathfinder.DIMENSION_NETHER;
+        else if (this.dimension == Level.END) dim = NetherPathfinder.DIMENSION_END;
+        else dim = NetherPathfinder.DIMENSION_END;
+        int height = Math.min(world.dimensionType().height(), 384);
+        if (!Baritone.settings().elytraAllowAboveRoof.value && dim == NetherPathfinder.DIMENSION_NETHER) height = Math.min(height, 128);
```
Ouch. Can you please autoformat your code with whatever formatting the rest of Baritone uses?

babbaj replied:
using properly formatted braces would make it take up more lines for no benefit

ZacSharp commented on src/main/java/baritone/process/elytra/NetherPathfinderContext.java:
```diff
@@ -24,98 +24,153 @@
+        else dim = NetherPathfinder.DIMENSION_END;
+        int height = Math.min(world.dimensionType().height(), 384);
+        if (!Baritone.settings().elytraAllowAboveRoof.value && dim == NetherPathfinder.DIMENSION_NETHER) height = Math.min(height, 128);
+        this.context = NetherPathfinder.newContext(seed, cache.toString(), dim, height, Baritone.settings().elytraCustomAllocator.value);
```
If `elytraUseBaritoneCache` is disabled then `cache` is null

ZacSharp commented on src/main/java/baritone/process/elytra/NetherPathfinderContext.java:
```diff
@@ -24,98 +24,153 @@
-                boolean isSolid = pair.second() != AIR_BLOCK_STATE;
-                Octree.setBlock(ptr, pos.getX() & 15, pos.getY(), pos.getZ() & 15, isSolid);
-            });
+            // not inserting or deleting from the cache hashmap so we should be fine not being exclusive
```
Except iirc C++ considers any data race undefined behavior and doing this across an ffi boundary won't make things better.

babbaj replied:
It's undefined behavior if another thread is writing to the data at the same time we are reading it but that is not the case here.

ZacSharp replied:
We are the writing thread here, but since we only have the read lock another thread may do the reading required to trigger undefined behavior.

ZacSharp replied:
In case you forgot / didn't see it, there's an `Octree.setBlock` call in this block. The hashmap reading is hopefully fine, as the comment says.

babbaj replied:
The pathfinder racing with this would be inconsequential however multiple calls to this to nearby blocks could cause bad data to be written so i will change it to use a write lock. The read executor is still single threaded so there currently isn't a difference anyways.

ZacSharp commented on src/main/java/baritone/process/elytra/NetherPathfinderContext.java:
```diff
@@ -24,98 +24,153 @@
-                writeChunkData(chunk, ptr);
+                writeLock.lock();
+                try {
+                    long ptr = NetherPathfinder.allocateAndInsertChunk(this.context, chunk.getPos().x, chunk.getPos().z);
```
I assume this always returns a completely zeroed chunk?

babbaj replied:
Yes

ZacSharp reviewed (changes requested):
Ok, didn't expect quite as much breakage, but this adds at least two ways to crash the game.
* Disabling `elytraUseBaritoneCache` unconditionally causes an NPE, crashing the game unless this happens inside a command
* Disabling `elytraAllowAboveRoof` while flying above the nether roof and then causing recalculation crashes the game. I didn't even get a crash report, but no segfault either?
* Enabling `elytraPredictTerrain` in the overworld leads to endless recalculation and chat spam. Actually not a crash though.

Some less severe points
* The wall of text printed by `#elytra` unless `elytraTermsAccepted` is now outdated. It mentions elytra flight only being for the nether (now wrong) and also recommends `elytraAutoJump`, which has hardly any chance of working in the overworld, and should also mention that terrain prediction is nether-only.
* The path is automatically recalculated when `elytraNetherSeed` changes, `elytraAllowThighsSpaces` does not have this effect. Not sure whether automatically recalculating is better.

I also really do not like how we are offsetting positions back and forth all across the code. Just look at how many times this went wrong with the existing chunk cache.
Can't we just use exactly one coordinate type (absolute world coordinates) in Baritone and shift to 0 based at the API boundary?
You didn't want to do this in nether-pathfinder itself, but maybe we can wrap the java side of the API to do the shifting?

babbaj commented:
> Disabling elytraUseBaritoneCache unconditionally causes an NPE, crashing the game unless this happens inside a command

fixed

> * Disabling `elytraAllowAboveRoof` while flying above the nether roof and then causing recalculation crashes the game. I didn't even get a crash report, but no segfault either?

I don't think this description is accurate. If elytraAllowAboveRoof is off and we put ourself above the roof then there is a crash (nether-pathfinder printing "retard" to stderr and calling std::terminate) but changing that setting while flying has no effect except possibly when it finds a landing spot.

I have not yet fixed this but I will probably fix it by pausing pathfinding until the player goes back below the roof.

> Enabling elytraPredictTerrain in the overworld leads to endless recalculation and chat spam. Actually not a crash though.

fixed

> The path is automatically recalculated when elytraNetherSeed changes, elytraAllowThighsSpaces does not have this effect. Not sure whether automatically recalculating is better.

fixed to behave the same

babbaj committed:
elytra fixes

babbaj committed:
hold write lock in queueBlockUpdate

ZacSharp commented:
You want to fix the out-of-bounds crash before merging this or later?

babbaj commented:
before. i also found another bug last night that's more important.

underscore-zi committed:
Keep zero-based coordinates to within pathfinder code

underscore-zi commented:
> I also really do not like how we are offsetting positions back and forth all across the code. Just look at how many times this went wrong with the existing chunk cache.
> Can't we just use exactly one coordinate type (absolute world coordinates) in Baritone and shift to 0 based at the API boundary?

I've added a commit on my branch (https://github.com/underscore-zi/baritone/commit/d0c3aeac783d8d97c859553849d1da8c888900db) that pulls all those adjustments into the  `NetherPathfinderContext` methods that make the pathfinder calls along with one Octree usage in `BlockStateOctreeInterface`.

On a related note, is there a better way for me to share commits? Or would you just prefer I left it all over to Babbaj now since its not my PR?

babbaj committed:
fix crashes

babbaj committed:
Merge remote-tracking branch 'origin/1.19.4' into nether-elytra-update

babbaj commented:
Just fixed the rest of the known bugs in https://github.com/cabaletta/baritone/pull/4698/commits/75096e1d352d0cc208d1f1be471ecf8f0e82259d

> I've added a commit on my branch (https://github.com/underscore-zi/baritone/commit/d0c3aeac783d8d97c859553849d1da8c888900db)

merged

babbaj committed:
final

babbaj committed:
rename elytraUseBaritoneCache

ZacSharp commented on src/main/java/baritone/process/elytra/ElytraBehavior.java:
```diff
@@ -159,9 +164,11 @@ public void tick() {
-            // Obstacles are more important than an incomplete path, handle those first.
-            this.pathfindAroundObstacles();
-            this.attemptNextSegment();
+            if (ctx.player().position().y < context.maxHeight && ctx.player().position().y > 0) {
```
This looks like comparing world coordinates against a range of zero based coordinates?

babbaj replied:
fixed

babbaj committed:
oops

ZacSharp commented:
Still get a crash by these steps
1. Stand on the nether roof with some firework rockets and an elytra
2. Run `#settings reset all`, `#elytraAllowAboveRoof true`, `#thisway 1000`, `#elytra`
3. Start flying
4. Run `#elytraAllowAboveRoof false`, `#elytra reset`

ZacSharp commented on src/main/java/baritone/process/elytra/ElytraBehavior.java:
```diff
@@ -1262,6 +1280,7 @@ private static Vec3 step(final Vec3 motion, final Vec3 lookDirection, final floa
         return new Vec3(motionX, motionY, motionZ);
     }
 
+    // any call to this must be done with the lock held
```
Can you maybe apply this comment to all methods which require synchronization and specify that the readlock is required?

babbaj replied:
This is the only function run on the main thread that requires synchronization but i added a comment to solveAngles.

ZacSharp replied:
As far as I can tell we have
```
passable called by simulate and pathFindAroundObstacles
simulate called by solvePitch called by the other solvePitch called by solveAngles
solveAngles called by onPostTick with synchronisation and by tick (and tick is public)
tick called by ElytraProcess.onTick with synchronization on the main thread
pathFindAroundObstacles called by PathManager.tick called by onTick0 called by onTick with synchronization on the main thread
```
At least `tick` needing synchronization should be documented.

ZacSharp commented on src/main/java/baritone/process/elytra/BlockStateOctreeInterface.java:
```diff
@@ -27,28 +28,31 @@ public final class BlockStateOctreeInterface {
+    public BlockStateOctreeInterface(final NetherPathfinderContext context, final DimensionType dimType) {
         this.context = context;
         this.contextPtr = context.context;
+        this.dimType = dimType;
```
Could potentially take the dimension info from `context`

babbaj replied:
NetherPathfinderContext just has the id of the dimension which the min y can't directly be taken from

ZacSharp replied:
It does have the min y here https://github.com/cabaletta/baritone/blob/04813a8312550ba5c18a0611b175f5062d21db46/src/main/java/baritone/process/elytra/NetherPathfinderContext.java#L83

ZacSharp commented:
> On a related note, is there a better way for me to share commits? Or would you just prefer I left it all over to Babbaj now since its not my PR?

You could open pull requests against babbaj's pull request branch, but I'm not sure whether that's better than a writing a comment.

babbaj committed:
fix #elytra reset bypassing y checks

babbaj committed:
i forgor to commit this

babbaj commented:
> Still get a crash by these steps

fixed

ZacSharp commented:
> > Still get a crash by these steps
> 
> fixed

Replace `#elytra reset` with changing `elytraNetherSeed`/`elytraPredictTerrain`/`elytraAllowTightSpaces` and it crashes once again.

babbaj committed:
fixes

babbaj commented:
fixed again

ZacSharp reviewed (approved):
I'm out of crashes now. The only thing I managed to find is an "Unhandled exception" error message when you change to an invalid goal mid-flight, but that's not new. (e.g. `#goal ~`).

leijurv commented:
I will merge this..... on one condition........

When you #elytra to a goal in the overworld that is at least some distance away (configurable setting, defaults to, let's say 1000 blocks), it should fly upwards (if it cannot fly upwards it'll go towards the goal of course). Once it gets above y=320, it will have a single target point that is simply the goal (ignoring the Y of the goal). I think this will work fine in the code because raytraces in minecraft have a maximum number of blocks they visit so it won't be too laggy? If it is laggy just add partial points at fractions along the way. Then once it is near the goal it dips back down below 320 and does landing or whatever at the goal

I think this would be a compromise that would allow it to fly through interesting overworld terrain like 2b2t spawn, while still going straight any-angle diagonal towards the goal (rather than the straight + 45° thing that I just think is low quality).

babbaj committed:
fix exception when setting out of bounds goal while elytra flying

babbaj commented:
It is currently not possible to fly above y 320 without increasing the height in nether-pathfinder even more. That should be an easy change.

babbaj committed:
fix deadlock

pizzav-xyz commented:
This is amazing

leijurv commented:
> It is currently not possible to fly above y 320 without increasing the height in nether-pathfinder even more. That should be an easy change.

No the whole point is to **not** use nether-pathfinder above the ceiling. When above y=320, you point in a straight line to the goal, without **any** pathfinding. Don't increase the height of nether-pathfinder's chunks

Have nether-pathfinder generate a path to y=320, then on the java side, add a path point after that that's at like same x,z but y=350 or something. Then the next path point is the goal x,z at y=350.

Might need to edit nether-pathfinder to support GoalYLevel

leijurv commented:
(and it should only do this if the destination is at least, e.g., 1000 blocks away. configurable setting)

underscore-zi commented:
> Have nether-pathfinder generate a path to y=320, then on the java side, add a path point after that that's at like same x,z but y=350 or something. Then the next path point is the goal x,z at y=350.

I don't have much time to work on this at the moment, but I originally used an implementation kinda like that and was working on a less hacked together implementation more recently but haven't finished. So I'm just kinda dumping what I had for my future-self or if someone else gives it a stab:

To allow it to follow a path that goes outside of the pathfinders limits `ElytraBehavior:clearView` has to be patched to use `ctx.world().clip(...)` instead of the native pathfinder when above the y-limit. And `ElytraBehavior:passable` to use `this.bsi` like the ignoreLava path instead of `this.boi`. 

Then for actually generating the path without pathfinder support for GoalY, I took advantage of the fact that it would attempt to pathFind again when near the end of an incomplete(finished=false) path.

When pathFinding it would check the source. If its above or near 320 and above the height map of the area it would just generate point path. If going far enough it would attempt to pathfind to 75%[0] the minimum distance with y=320, or it would just pathfind to the destination.

This allowed it to fly under a roof by just repeatedly creating incomplete paths going 75% the minimum distance and eventually switch to the point path whenever the route became clear.

Also its worth noting that the point path did need to be more detailed than just the destination point as it didn't like to fly if the next point wasn't loaded in. Though that might be fixable by patching `clearView` to be true when at elevation instead of using `clip(...)`. 

[0] 75% wasn't chosen for any particular reason, just giving some breathing room

CREAsTIVE commented:
any updates?

underscore-zi committed:
Create an interface for elytra pathfinding

underscore-zi committed:
Add IElytraContextFactory

lawnguy1201 commented:
What has happened with this?

pizzav-xyz commented:
> What has happened with this? 
> 

A random guy has commented on this 3 months later

lawnguy1201 commented:
> > What has happened with this?
> 
> A random guy has commented on this 3 months later

no shit

pizzav-xyz commented:
> > > What has happened with this?
> > 
> > A random guy has commented on this 3 months later
> 
> no shit

So why are you asking if you already know

underscore-zi commented:
Still actively working on it.

 I have something that generally works and goes above the build limit. Though I've been working out some bugs where it seems to get stuck momentarily as if it can't see the next node (despite clear raytraces to it).

Also some issues with generated the transitions between the different path generators since it usually tries to transition right at the edge of loaded chunks when it tries to pathfind right when a new chunk gets loaded.

lawnguy1201 commented:
> Still actively working on it.
> 
> I have something that generally works and goes above the build limit. Though I've been working out some bugs where it seems to get stuck momentarily as if it can't see the next node (despite clear raytraces to it).
> 
> Also some issues with generated the transitions between the different path generators since it usually tries to transition right at the edge of loaded chunks when it tries to pathfind right when a new chunk gets loaded.

Nice, would there be anyway I could get the .jar. working on a project and it need B to fly in the OW.

underscore-zi commented:
I'm trying to troubleshoot some behavior and wondering if any of you all might be able to point me to where in code this behavior might be coming from.

[2025-08-22-113218482-2c80a.webm](https://github.com/user-attachments/assets/a48c0508-7bb9-4cd6-8c59-e18cd3ec2c15)

Basically, as the Elytra crosses the build limit it starts to have trouble following the path. It'll follow the path, then seem to get stuck and start spinning. This is similar to the behavior that happens when about to enter an unloaded chunk though it is loaded in this case. 

I suspect something sees it as unloaded (`ctx.world().isLoaded` returns false for y > build limit, but I've patched that already in `ElytraBehavior`). Though I'm not sure where this spin behavior comes from exactly. And if I reduce the flight level to be under the build limit, it works without issue. So I'm thinking there is some check buried away I've missed besides the one in `attemptNextSegment`

Edit: and for what its worth raytraces all come up green.

pizzav-xyz commented:
> I'm trying to troubleshoot some behavior and wondering if any of you all might be able to point me to where in code this behavior might be coming from.
> 2025-08-22-113218482-2c80a.webm
> 
> Basically, as the Elytra crosses the build limit it starts to have trouble following the path. It'll follow the path, then seem to get stuck and start spinning. This is similar to the behavior that happens when about to enter an unloaded chunk though it is loaded in this case.
> 
> I suspect something sees it as unloaded (`ctx.world().isLoaded` returns false for y > build limit, but I've patched that already in `ElytraBehavior`). Though I'm not sure where this spin behavior comes from exactly. And if I reduce the flight level to be under the build limit, it works without issue. So I'm thinking there is some check buried away I've missed besides the one in `attemptNextSegment`
> 
> Edit: and for what its worth raytraces all come up green.

are you sure the chunks arent actually unloaded

underscore-zi commented:
> are you sure the chunks arent actually unloaded

Yes

ZacSharp commented:
Having a short look at the solver it looks like `isHitboxClear` might also need patching to not use `context.raytrace`. Not sure why `passable` needs to use `bsi` instead of `boi` since `boi.get` returns `false`/air for out of bounds positions, which is what we need here.

leijurv commented:
yeah looks like something messed up with the raytracer. out of bounds (above the ceiling) should return air

underscore-zi commented:
Thanks! Fixed it. Was able to trace the bug down from `isHitBoxClear`.

It was indeed a raytracing bug, miscalculated when a target was in/out of bounds resulting in misusing the NetherPathfinder for OOB raytraces.

nonokirby commented:
Is this pull stable enough to fork and run currently, or does it have major issues still? specifically looking to elytrafly in/around 2b's spawn if knowing terrain helps.

ZacSharp commented:
As far as I know there are no known (new) bugs, so should definitely be usable.

(The reason this isn't merged is a pending functional change)

underscore-zi committed:
Implementation of flight above build limit for long distances

underscore-zi committed:
Don't use the Sky pathfinder when in the nether and no roof allowed

underscore-zi commented:
I have some commits implementing flight above the build limit over on: https://github.com/underscore-zi/baritone 

I do make one major structural change which might raise some eyebrows. I've extended the API with `IElytraPathfinderContext` and `IElytraPathfinderContextFactory` 

So now, the NetherPathfinderContext basically already implemented the interface, and I've added a new SkyPathfinderContext which also implements the interface and wraps the NetherPathfinder to extend it adding support for flight above the build limit. ElytraBehavior uses the appropriate context for the dimension/settings, keeping the original performance if users want that.

In defense of having this in the API is does serve to implement a feature I've wanted, the ability to navigate between waypoints without doing the whole takeoff/fly/land process for every point or changing the destination when the flight gets close to each point. I made a nice like flight replayer for myself with this.

underscore-zi committed:
Prevent races between destruction and creation of ElytraBehavior

underscore-zi committed:
Prevent deadlock when rapidly changing goals and the solver never gets executed/finishes by adding a timeout

ZacSharp commented:
Tried to come up with a funky way to review this. Wouldn't have expected that two people working on one pr is such a big problem on a "collaboration platform".

`SkyPathfinderContext` uses the wrong naming convention for it's private helpers.

https://github.com/underscore-zi/baritone/compare/20c1631abfe8b64fa119e8012b348ebdd55a32a8...94a466be9018e87e3f55e9e5148cc32b4ed5e608#diff-aa0f9f28257c586978349ec09d0ec52c19858ebfc79ff63b56b498a69ca16d14R28
I don't like exposing nether-pathfinder API directly in our API. And `UnpackedSegmend` is imho a prime example of what's considered an implementation detail.

https://github.com/underscore-zi/baritone/compare/20c1631abfe8b64fa119e8012b348ebdd55a32a8...94a466be9018e87e3f55e9e5148cc32b4ed5e608#diff-f881174eae586d93b19f8eb3f9c26af571c0ca278f7d9a108cba4d66a71f519fR397-R398
Aren't there fields caching these two values?


https://github.com/underscore-zi/baritone/compare/20c1631abfe8b64fa119e8012b348ebdd55a32a8...94a466be9018e87e3f55e9e5148cc32b4ed5e608#diff-f881174eae586d93b19f8eb3f9c26af571c0ca278f7d9a108cba4d66a71f519fR443-R458
It looks weird to have both a special case when the field is null in the getter and a null-replacement in the setter.


https://github.com/underscore-zi/baritone/compare/20c1631abfe8b64fa119e8012b348ebdd55a32a8...94a466be9018e87e3f55e9e5148cc32b4ed5e608#diff-eebf5e50d8fe55d39379412b51279851bba3a6f3019584be8ea2bfd36ce71fe1R85
I like how this is no longer being passed around. It uses the exact same internal storage and is fundamentally tied to the context, so might as well be part of it.

<br>

Honestly I'm not happy with the two new interfaces. `IElytraPathfinderContextFactory` feels like it's mostly boilerplate and `IElytraPathfinderContext` is just bad API. It is what netherpathfinder expects and using it internally makes sense, but it's a messy implementation detail.
There's also a concern about deadlocks here, because giving API users control of the context lock means the thing can lock and unlock in ways we do not expect. The thread "safety" in this part of Baritone is a mess already.

Flying around without landing sounds very reasonable though, so maybe there's a way to implement that feature without putting implementation details into the API package?

<br>

A somewhat unrelated idea I had is that iirc the pathfinder uses a symmetric movement set so there is no difference between going forwards or backwards. That means if the goal is already in render distance and we are still above the world we could run the pathfinder from the goal towards the build limit / our current position to determine where to transition down. Not sure whether that helps with anything.

babbaj committed:
Merge remote-tracking branch 'origin/1.19.4' into nether-elytra-update

# Conflicts:
#	src/main/java/baritone/process/ElytraProcess.java
#	src/main/java/baritone/process/elytra/ElytraBehavior.java
#	src/main/java/baritone/process/elytra/NetherPathfinderContext.java

babbaj commented:
just merged his branch into mine. he was also a few commits behind so I had to merge and fix some conflicts.

underscore-zi commented:
> And `UnpackedSegmend` is imho a prime example of what's considered an implementation detail.

I feel like the `Stream<BetterBLockPos>` and boolean wrapper are what would be necessary to expose to allow someone else to provide a path. The name is a bit off though since the public side would have no knowledge of the "packed" segments. 

Perhaps a simpler option would be to only allow the API user to provide waypoints and leave the path to Baritone. That perhaps fits better with Baritone more generally and would probably be the main usecase. One use-case I had for full pathing was flying down 2b diagonal highways where there was some active wither griefing so I couldn't bounce.

> I don't like exposing nether-pathfinder API directly in our API.

Yeah, I wasn't the biggest fan of my own approach there though I mostly was just trying not to break anything outside of my own code. So isolated it via the abstraction. But yeah, I think I can clean that up quite a bit.

Raytracing stuff can be pulled back internal, no reason to have that overloadable if the base implementation supports OOB traces it handles all cases. Though it adds a bit of overhead as it needs to check which tracer to use. Chunk/block information can be obtained by any mod already so no point in providing that either; that also removes the locks.

Which would leave pathFindAsync, cancel and maybe destroy which feels much more sane. And I think that might remove the Factory too as there wouldn't be a need to provide the Baritone cache path into it.

underscore-zi commented:
Committed to my fork again: https://github.com/underscore-zi/baritone

I pulled everything new out of the API, though I still use a simpler interface internally so the original Nether code can be swapped in depending on settings.

Murat65536 commented:
Fixing a few of the things Codacy is mad about might also be a good idea. Mostly naming things. I wouldn't worry about the NPath complexity.

underscore-zi commented:
I believe I've addressed all of the issues from Codacy that were still relevant. 👍

ZacSharp commented:
Looks better now, thanks.

Feeding a set of waypoints sounds good (honestly for the normal pathfinder as well, but that can be kept separate).

underscore-zi commented:
I was thinking to leave waypoints out of this PR and I'll make that a separate one for that.

I do have one bug to fix in this set though. The semaphore I introduced leads (I suspect it's the cause atleast) to a bunch of lag when it gets stuck and can't find a path forward and keeps attempting to recalculate.

ZacSharp commented:
> I was thinking to leave waypoints out of this PR and I'll make that a separate one for that.

:+1:

underscore-zi commented:
I pushed three more commits

1. First was just a bug when it no landing spot was found, instead of hovering ti would land on whatever block was there. Not sure exactly when/how that behavior changed but easily restored.
2. I thought the lag was due to the semaphore and some path adjustments leading to creation/destroying of NetherPathfinderContext with new pathing requests. So it pulls the context object out of the Behavior allowing it to persist on new pathTo request if the context already exists. Ultimately this commit might not be necessary as it wasn't the issue, but I didn't revert it as performance-wise it still makes sense to not need to destroy and recreate the chunk cache every time especially with use-cases like Jeff Mod's trail follower which frequently sets new pathing goals.
3. Actual issue was just that the PathCalculationException was getting wrapped as a RuntimeException which once propagated up led to a bunch of extra logging and handling and they would happen in rapid succession.

kaliko69 commented:
Hey! I’m really excited about the Elytra auto-flight in the Overworld/End feature. I tried versions 1.21.4 and 1.21.8, but it seems like only Nether Elytra works there. Is there a way to get the Overworld/End Elytra version right now, or is it still work-in-progress? Thanks! ( i dont know who to mention @babbaj @underscore-zi  @ZacSharp @leijurv  @kittenvr )

underscore-zi commented:
@babbaj would you mind pulling in my commits? I believe its good to go now; though I'll address any new issues from codacy.

leijurv commented:
ignore codacy

ZacSharp commented:
@babbaj by GitHub permissions logic I have push access to this pr so if you don't mind I can also transfer zi's commits here.

dep4 commented:
Can this be merged, pretty please?
